### PR TITLE
RFC: add OS.MM

### PR DESCRIPTION
**Pull request by talex5**
Provides a `Heap_pages` module which provides access to the size of the heap and the current number of free pages. This is useful for detecting memory pressure.

### CODE REVIEW

**Comment by avsm** [comment]
could we have a unix equivalent of this? would help with benchmark suites.

File: bindings/mm_stubs.c [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+#include <caml/mlvalues.h>
+#include <mini-os/mm.h>
+
+CAMLprim value
```
Comment (avsm) [inline bindings/mm_stubs.c]: noalloc?

Comment (talex5) [inline bindings/mm_stubs.c]: Good idea - done.

**Comment by talex5** [comment]
@avsm What would the Unix version do? There's usually effectively unlimited memory available there. Could check rlimits, although cgroups might be more common, on Linux at least.

**Comment by avsm** [comment]
I was thinking of it primarily as a convenient way to record memory usage in test scripts while developing. The general "shape" of the heap usage should be very similar under Unix and Xen, and measuring it is generally easier from a Unix userspace than running under Xen.

**Comment by talex5** [comment]
Ah, so providing `used` but not `total` on Unix? What should it return? `rusage(2)`'s `ru_idrss` looks closest, but man-page says not supported on Linux.

On Xen it's very simple: it's the number of pages currently allocated by the page allocator, out of the fixed heap allocated at startup based on the VM's memory allocation.

**Comment by talex5** [comment]
Can this be merged now, or does someone have a better plan? Maybe we should provide this as the low-level interface and provide some higher-level abstraction elsewhere?

**Comment by avsm** [comment]
This was merged but not released...

**Pull request merged**